### PR TITLE
[Feature] Improve K8S API Operator First-Hand Experience

### PR DESCRIPTION
**Pull request by renuka-fernando**
Fix wso2/k8s-apim-operator#184

#### Add command `apictl install api-operator` with the flag `-f`
- Read registry specific inputs
- Install OLM and api-operator

#### Introduce type: `registry`
```go
// Registry represents Docker Registry
type Registry struct {
	Name       string  // Unique Name
	Caption    string  // Text to display in the CLI about registry details
	Repository *string // Repository name
	Option     int     // Option to be choose the CLI registry list
	Read       func()  // Function to be called when getting inputs
	Run        func()  // Function to be called when updating k8s secrets
}
```

#### Supported registry types
- Private Docker-Hub
- Private HTTP/HTTPS registries
- Amazon ECR (with Amazon EKS)
- GCR
- Quay.io

#### Others
- Read API operator and OLM versions from OS Environment variables
- `change registry` command to change the registry type

### CODE REVIEW

File: import-export-cli/sonar-project.properties [hunk 1]
```diff
@@ -0,0 +1,9 @@
+sonar.projectKey=com.wso2.import-export-cli
+sonar.projectName=Import Export CLI (API-CTL)
+
+sonar.sources=.
+sonar.exclusions=**/*_test.go,**/vendor/**
+
+sonar.tests=.
+sonar.test.inclusions=**/*_test.go
+sonar.test.exclusions=**/vendor/**
```
Comment (HiranyaKavishani) [inline import-export-cli/sonar-project.properties:9]: Need an extra line EOF

File: import-export-cli/cmd/changeDockerRegistry.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/cmd/changeDockerRegistry.go:9]: Align with line

File: import-export-cli/cmd/install.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/cmd/install.go:9]: same here..

File: import-export-cli/cmd/installApiOperator.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/cmd/installApiOperator.go:9]: Align with line

File: import-export-cli/cmd/uninstall.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/cmd/uninstall.go:9]: same here

File: import-export-cli/cmd/uninstallApiOperator.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/cmd/uninstallApiOperator.go:9]: same here

File: import-export-cli/operator/olm/consts.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/olm/consts.go:9]: same here

File: import-export-cli/operator/olm/install.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/olm/install.go:9]: remove space

File: import-export-cli/operator/registry/amazon_ecr.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
+package registry
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/registry/amazon_ecr.go:1]: License header is missing

File: import-export-cli/operator/registry/dockerhub.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/registry/dockerhub.go:9]: Remove space

File: import-export-cli/operator/registry/gcr.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/registry/gcr.go:9]: remove space

File: import-export-cli/operator/registry/http.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/registry/http.go:9]: remove space

File: import-export-cli/operator/registry/registry.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
+/*
+*  Copyright (c) WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (HiranyaKavishani) [inline import-export-cli/operator/registry/registry.go:9]: same here

File: import-export-cli/utils/httpUtils.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package utils
```
Comment (hisanhunais) [inline import-export-cli/utils/httpUtils.go:1]: Shall we add the License Header here?

File: import-export-cli/operator/registry/registry.go [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+	// apply controller config config map back
+	if err := k8sUtils.K8sApplyFromStdin(string(configuredConfigMap)); err != nil {
+		utils.HandleErrorAndExit("Error creating controller-configs", err)
+	}
+}
+
+// add adds a registry to the registries maps
+func add(registry *Registry) {
+	if registry.Option < 1 {
+		utils.HandleErrorAndExit("Error adding registry: "+registry.Name, errors.New("'option' should be positive"))
```
Comment (hisanhunais) [inline import-export-cli/operator/registry/registry.go:116]: Shall we add a space before and after the '+'?

File: import-export-cli/operator/registry/registry.go [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	if registries[registry.Option] != nil {
+		utils.HandleErrorAndExit("Error adding registry"+registry.Name, errors.New("duplicate 'options' values"))
```
Comment (hisanhunais) [inline import-export-cli/operator/registry/registry.go:119]: Shall we add a space before and after the '+'?

File: import-export-cli/operator/registry/dockerhub.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+	Read: func() {
+		repository, username, password := readDockerHubInputs()
+
+		dockerHubValues.repositoryUrl = getRegistryUrl(repository)
+		dockerHubValues.username = username
+		dockerHubValues.password = password
+
+		// Docker Hub not supports "docker.io/foo" hence make repository as "foo"
+		if isDockerHub(repository) {
+			repository = strings.TrimPrefix(repository, DockerHubInputPrefix+"/")
```
Comment (hisanhunais) [inline import-export-cli/operator/registry/dockerhub.go:58]: Shall we add a space before and after the '+'?

File: import-export-cli/cmd/uninstallApiOperator.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+			fmt.Println("Cancelled operation: uninstalling API Operator")
+		}
+
+	},
+}
+
+func init() {
+	uninstallCmd.AddCommand(uninstallApiOperatorCmd)
+	uninstallApiOperatorCmd.Flags().BoolVar(&flagForceCleanApiOperator, "force", false, "Force uninstall API Operator")
+	//uninstallApiOperatorCmd.Flags().BoolVar(&flagCleanOlmOperator, "remove-olm", false, "Clean kubernetes resources for OLM Operator as well")
```
Comment (hisanhunais) [inline import-export-cli/cmd/uninstallApiOperator.go:109]: Shall we remove the commented code?

File: import-export-cli/cmd/uninstallApiOperator.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+			// delete the OLM subscription
+			fmt.Println("Removing OLM subscription")
+			_ = k8sUtils.ExecuteCommand(k8sUtils.Kubectl, k8sUtils.K8sDelete, "subscriptions.operators.coreos.com", "my-api-operator", "-n", "operators")
+
+			// delete namespace wso2-system
+			//fmt.Printf("Removing namespace: %s\nThis operation will take some minutes...\n", k8sUtils.ApiOpWso2Namespace)
+
+			deleteErrors := []error{
+				//k8sUtils.ExecuteCommand(k8sUtils.Kubectl, k8sUtils.K8sDelete, "namespace", k8sUtils.ApiOpWso2Namespace),
```
Comment (hisanhunais) [inline import-export-cli/cmd/uninstallApiOperator.go:86]: Shall we remove the commented code?

File: import-export-cli/cmd/uninstallApiOperator.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				utils.HandleErrorAndExit("Error in API Operator version", err)
+			}
+
+			// delete the OLM subscription
+			fmt.Println("Removing OLM subscription")
+			_ = k8sUtils.ExecuteCommand(k8sUtils.Kubectl, k8sUtils.K8sDelete, "subscriptions.operators.coreos.com", "my-api-operator", "-n", "operators")
+
+			// delete namespace wso2-system
+			//fmt.Printf("Removing namespace: %s\nThis operation will take some minutes...\n", k8sUtils.ApiOpWso2Namespace)
```
Comment (hisanhunais) [inline import-export-cli/cmd/uninstallApiOperator.go:83]: Shall we remove the commented code?

File: import-export-cli/cmd/uninstallApiOperator.go [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+const uninstallApiOperatorCmdLiteral = "api-operator"
+const uninstallApiOperatorCmdShortDesc = "Uninstall API Operator"
+const uninstallApiOperatorCmdLongDesc = "Uninstall API Operator in the configured K8s cluster"
+const uninstallApiOperatorCmdExamples = utils.ProjectName + ` ` + uninstallCmdLiteral + ` ` + uninstallApiOperatorCmdLiteral + `
+` + utils.ProjectName + ` ` + installApiOperatorCmdLiteral + ` --force`
+
+var flagForceCleanApiOperator bool
+
+//var flagCleanOlmOperator bool
```
Comment (hisanhunais) [inline import-export-cli/cmd/uninstallApiOperator.go:37]: Shall we remove the commented code?

**Pull request merged**